display_list/display_list_canvas_unittests.cc

```diff
@@ -229,106 +238,209 @@ class BoundsTolerance {
   SkPoint scale_ = {1, 1};
   SkPoint absolute_pad_ = {0, 0};
   SkRect clip_ = {-1E9, -1E9, 1E9, 1E9};
+  SkPoint clip_pad_ = {0, 0};
 
   SkScalar discrete_offset_ = 0;
 };
 
 typedef const std::function<void(SkCanvas*, SkPaint&)> CvSetup;
 typedef const std::function<void(SkCanvas*, const SkPaint&)> CvRenderer;
 typedef const std::function<void(DisplayListBuilder&)> DlRenderer;
-static void EmptyCvRenderer(SkCanvas*, const SkPaint&) {}
-static void EmptyDlRenderer(DisplayListBuilder&) {}
+static const CvSetup kEmptyCvSetup = [](SkCanvas*, SkPaint&) {};
+static const CvRenderer kEmptyCvRenderer = [](SkCanvas*, const SkPaint&) {};
+static const DlRenderer kEmptyDlRenderer = [](DisplayListBuilder&) {};
+
+using PixelFormat = DlSurfaceProvider::PixelFormat;
+using BackendType = DlSurfaceProvider::BackendType;
 
-class RenderSurface {
+class RenderResult {
  public:
-  explicit RenderSurface(sk_sp<SkSurface> surface)
-      : surface_(std::move(surface)) {
-    EXPECT_EQ(canvas()->save(), 1);
-  }
-  ~RenderSurface() { sk_free(addr_); }
-
-  SkCanvas* canvas() { return surface_->getCanvas(); }
-
-  const SkPixmap* pixmap() {
-    if (!pixmap_.addr()) {
-      canvas()->restoreToCount(1);
-      SkImageInfo info = surface_->imageInfo();
-      if (info.colorType() != kN32_SkColorType ||
-          !surface_->peekPixels(&pixmap_)) {
-        info = SkImageInfo::MakeN32Premul(info.dimensions());
-        addr_ = malloc(info.computeMinByteSize() * info.height());
-        pixmap_.reset(info, addr_, info.minRowBytes());
-        EXPECT_TRUE(surface_->readPixels(pixmap_, 0, 0));
-      }
-    }
-    return &pixmap_;
+  RenderResult(const sk_sp<SkSurface>& surface,
+               const SkMatrix& matrix,
+               const SkIRect& clip_bounds)
+      : matrix_(matrix), clip_bounds_(clip_bounds) {
+    SkImageInfo info = surface->imageInfo();
+    info = SkImageInfo::MakeN32Premul(info.dimensions());
+    addr_ = malloc(info.computeMinByteSize() * info.height());
+    pixmap_.reset(info, addr_, info.minRowBytes());
+    EXPECT_TRUE(surface->readPixels(pixmap_, 0, 0));
   }
+  ~RenderResult() { free(addr_); }
+
+  int width() const { return pixmap_.width(); }
+  int height() const { return pixmap_.height(); }
+  const SkMatrix& matrix() const { return matrix_; }
+  const SkIRect& clip_bounds() const { return clip_bounds_; }
+  const uint32_t* addr32(int x, int y) const { return pixmap_.addr32(x, y); }
 
  private:
-  sk_sp<SkSurface> surface_;
   SkPixmap pixmap_;
   void* addr_ = nullptr;
+  SkMatrix matrix_;
+  SkIRect clip_bounds_;
+};
+
+#define RENDER_JOB_BASE                 \
```

Comment:
Why couldn't this just be a base struct struct?

Reply:
I wanted to use struct initializers throughout at the points of invocation, but that style of initialization cannot work with a v-table. So, no virtual methods -> no struct initializers -> complicated multi-argument constructors with arguments that are only used in some cases -> harder to read invocation code.

Reply:
I turned this into 2 structs - non-virtual `RenderJobInfo` with the optional attributes and named struct initialization, and a virtual `JobRenderer` which hosts the necessary Render() method, along with some specific methods used to retrieve task-specific info during the testing. No more templates or macros.